### ISSUE TITLE
bug: Form instances are not Garbage Collected and neither its containing modules as a side effect.

### ISSUE THREAD

**Issue by burdokat** [ionic-team/ionic#21521]
# Bug Report

**Ionic version:**

[ ] **4.x**
[x] **5.x**

**Angular version:**
[x] **9.x**

**Current behavior:**

When navigating into and out of a page containing a template or reactive driven form, the instances pertaining the form are not garbage collected. Hence, every input field data, input instances, validators, etc, and everything else inside the form container (page, component, or modal), are not garbage collected when navigating out of the page.

**Expected behavior:**

Any instance pertaining the form and it's container must be garbage collected when navigating out.

**Steps to reproduce:**

* Follow the link to this repository:
https://github.com/burdokat/FormDemo

**Related code:**
* Follow the link to this repository:
https://github.com/burdokat/FormDemo

**Additional Information:**
* For more information and details of this bug, please follow the link to the repository below and check the README file:
https://github.com/burdokat/FormDemo
May be related to other bugs (requires more thorough investigation).
Help needed.

**Ionic info:**

```
Ionic:

   Ionic CLI                     : 6.9.1 (C:\Users\BRDKT\AppData\Roaming\npm\node_modules\@ionic\cli)
   Ionic Framework               : @ionic/angular 5.2.1
   @angular-devkit/build-angular : 0.901.8
   @angular-devkit/schematics    : 9.1.8
   @angular/cli                  : 9.1.8
   @ionic/angular-toolkit        : 2.2.0

Utility:

   cordova-res : not installed
   native-run  : not installed

System:

   NodeJS : v12.16.3 (C:\Program Files (x86)\nodejs\node.exe)
   npm    : 6.14.4
   OS     : Windows 8.1

```

**Comment by liamdebeasi** [comment]
Thanks for the issue. I am going to close this as this is not a bug in Ionic Framework.

When you do `this.router.navigate(['explore'])`. Angular Routing pushes a new page onto the stack. The previous page is still in the stack, in the event users want to go back to it. As a result, this prevents the page from being garbage collected.

It looks like you do not want this, so you should use the additional options Angular Routing provides: https://angular.io/api/router/NavigationExtras. Two good options here would be:

1. `skipLocationChange`. This navigates without pushing a new state into history.
2. `replaceUrl`. This navigates while replacing the current state in history.

Since you are going to/from a login page, it probably makes sense to ensure that users cannot accidentally "go back" to the page unless they explicitly logged out. Additionally, if a user logs out it probably makes sense to ensure that users cannot "go back" and access data from when they were logged in.

Hope this helps!

**Comment by burdokat** [comment]
@liamdebeasi I don't understand.
What is not working properly is the reactive and the template forms.... I strongly suggest you look inside the repository I uploaded. The pages and the observables are memory leaking and it is because of the form instances not being garbage collected. That's the issue in question.

**Comment by liamdebeasi** [comment]
Thanks for the follow up. I did look at your repository.

Your application starts on the "Explore" page. On app load, the page instance is initialized and added to the navigation stack. When you click "Logout", an "Auth" page instance is initialized and added to the navigation stack. This includes initializing forms, components, etc.

When you click the "Login" button on the "Auth" page, the way you are navigating is causing _another_ instance of the "Explore" page to be created. When you click the "Logout" button on that page, _another_ instance of the "Auth" page is being created. This includes creating another instance of the form, and adding the page to the navigation stack.

In other words, the way you are doing navigation is causing multiple instances of the "Auth" page to be created that are all being held in memory. As a result, the browser cannot garbage collect these pages since the user may still navigate back to them.

Using the suggestions in https://github.com/ionic-team/ionic/issues/21521#issuecomment-644133853 would remedy this behavior.

**Comment by burdokat** [comment]
@liamdebeasi Please, do please read and properly test my repository before answering, please... Did you heap snapshot the thread... did you searched the ngModel instances?? did you searched the observable instances??... did you navigate without touching the form???.... Now, do please navigate without touching the form. In other words, enable the login submit button (do not touch the input form) and navigate back and forth just pressing the buttons, and you will see the expected behaviour and you will see that the router is fine.... Whenever you touch the form, there is memory leaking and instances piles up like crazy. As I said before, it has nothing to do with the router.

**Comment by liamdebeasi** [comment]
Yes, I did thoroughly test your repository. This is not a bug in Ionic Framework, this is how routing works.

When you navigate from page A to page B, you are not destroying page A since users may want to navigate back to page A. What this means is that page A is stored in the navigation stack and is _not_ garbage collected. This is the default behavior of routing.

Your application is creating new instances of the Login page every time you navigate to it. Since the old login pages are stored in the navigation stack, they are not garbage collected. Again, this is not a bug, this is how routing works by default.

As per my comment in https://github.com/ionic-team/ionic/issues/21521#issuecomment-644133853 I recommend you do something like `this.router.navigate(['explore'], { replaceUrl: true });` when navigating, as it will replace the previous state and should allow you to avoid creating multiple instances of the same page.

You can also try using the `NavController` in Ionic Angular to set the root page:

```typescript
import { NavController } from '@ionic/angular';

...

constructor(private navCtrl: NavController) {}

...

goToPage() {
  this.navCtrl.navigateRoot('/explore');
}
```

This will clear out the navigation stack and set the `explore` page as the root page in the stack. This should allow old form instances to be garbage collected as well.

**Comment by burdokat** [comment]
You are not listening... Please navigate without touching the form. And you will see that is has nothing to do with the router...
You are the worst, I don't want you to tell me what to do. I am trying to explain to you how this bug is behaving. Please navigate without touching the form.

**Comment by liamdebeasi** [comment]
Let's not get into any personal insults as that will not help anyone. I am not telling you what to do, I am merely offering suggestions as to how you can get around the behavior you are experiencing. Whether or not you implement those suggestions is entirely up to you.

Can you please clarify what you mean by "navigate without touching the form"?

**Comment by burdokat** [comment]
@liamdebeasi enable the submit button inside the login page (delete the disable property), and just navigate back and forth without touching the inputs in the form. Just navigate pressing the buttons.

Please don't mind my rudeness if I was.
I took my time to open another thread with screenshots to further explain this issue... I think I have not been understood, so I revised my report and I hope this new thread could clarify what I am trying to describe as the issue. I hope this could help get me through:

https://github.com/ionic-team/ionic/issues/21539
